File: byteback-test/test-scala-2.13.8/src/main/scala/byteback/test/exceptions/MultiCatch.scala
/**
  * RUN: %{byteback} -cp %{jar} -c %{class} -o %t.bpl
  */
package byteback.test.exceptions;

import byteback.specification.Contract._;
import byteback.specification.Special._;
import byteback.specification.Operators._;
import byteback.specification.Operators.{eq => equal};

import scala.annotation.meta._;

class MultiCatch {

  @Exceptional
  @Behavior
  def always_throws_exception1_exception2(): Boolean = {
    return thrown().isInstanceOf[Exception1] | thrown().isInstanceOf[Exception2]
  }

  @Ensure("always_throws_exception1_exception2")
  def alwaysThrowsMultiple(): Unit = {
    throw new Exception1();
  }

  def emptyMulticatch(): Unit = {
    try {
      alwaysThrowsMultiple();
    } catch {
      case e @ (_: Exception1 | _: Exception2) =>
    }
  }

  def multiCatchUnionAssertion(): Unit = {
    try {
      alwaysThrowsMultiple();
      assertion(false);
    } catch {
      case e @ (_: Exception1 | _: Exception2) =>
        assertion(e.isInstanceOf[Exception1] | e.isInstanceOf[Exception2]);
    }
  }

}
/**
  * RUN: %{verify} %t.bpl | filecheck %s
  * CHECK: Boogie program verifier finished with 4 verified, 0 errors
  */

File: byteback-test/test-java-8/src/main/java/byteback/test/exceptions/MultiCatch.java
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} -o %t.bpl
 */
package byteback.test.exceptions;

import static byteback.specification.Contract.*;

public class MultiCatch {

	@Exceptional
	@Behavior
	public boolean always_throws_exception1_exception2() {
		return thrown() instanceof Exception1;
	}

	@Ensure("always_throws_exception1_exception2")
	public void alwaysThrowsMultiple() throws Exception {
		throw new Exception1();
	}

	public void emptyMulticatch() throws Exception {
		try {
			alwaysThrowsMultiple();
		} catch (Exception1 | Exception2 e) {
		}
	}

	public void multiCatchUnion() throws Exception {
		try {
			alwaysThrowsMultiple();
			assertion(false);
		} catch (Exception1 | Exception2 e) {
			assertion(e instanceof Exception1 | e instanceof Exception2);
		}
	}

	public void multiCatchFinally() throws Exception {
		try {
			alwaysThrowsMultiple();
			assertion(false);
		} catch (Exception3 | Exception4 e) {
			assertion(false);
		} finally {
		}
	}

}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 5 verified
 */
